[Fleet][Kafka Output] Tweaks for Kafka UI/API

szwarckonrad opened the pull request:
1. Removed Kerberos option from auth type selection
2. Used EuiFieldPassword component for password input
State:
![Screenshot 2023-08-01 at 11 16 30](https://github.com/elastic/kibana/assets/29123534/3ef0b3f8-c98b-42d0-9645-fdb6437859f5)
3. Added `None` authentication option. This will save `auth_type` as `none` and won't pass `usernamer`, `password` nor `ssl` to the SO.
![Screenshot 2023-08-02 at 15 59 08](https://github.com/elastic/kibana/assets/29123534/9bbe9ce7-2749-484f-833d-72d1c9e5ee46)
4. Renamed `broker_ack_reliability` to `required_acks` in UI and API. Changed value type to `1 || 0 || -1`
5. Renamed `broker_buffer_size` to `channel_buffer_size` in UI and API.
6. Temporarily disabled `password` field encryption.
7. `partition` key stored in yaml follows proper requirements
8. `topics` key stored in yaml follows proper requirements

kilfoyle commented:
Hi @szwarckonrad! I'm working on a [docs PR](https://github.com/elastic/ingest-docs/pull/336) for these UI changes. 

I have a couple of suggestions if you don't mind:

1. It might be good to have the Kafka help text match what we have for the Logstash output settings UI:
    Kafka output settings UI:
    ![kafka-ui-text](https://github.com/elastic/kibana/assets/41695641/8e480a02-b974-4338-bf0b-3f3f6eb19335)
    Logstash output settings UI:
    ![logstash-ui-text](https://github.com/elastic/kibana/assets/41695641/4ec09931-d635-46de-bac7-c51f830a5496)
   Could we change the Kafka UI text to `Specify the URLs that your agents will use to connect to Kafka. Learn more`?

2. Once merged, my docs PR will create individual pages for the Logstash output settings and the Kafka output settings. For those "Learn more" links, would it be possible to have them point to `<base-path>/ls-output-settings.html` and `<base-path>/kafka-output-settings.html`, respectively, instead of the main UI settings page (`fleet-settings.html`)? 

Thanks!

kevinlog commented:
@szwarckonrad Thank you for putting together these changes, it's a massive help!

I was chatting with @faec and there are a couple more changes in the conversion to `elastic-agent.yml` that we should tweak:

specifically `partition` and `topics`

#### Partition

We need to follow the rules here: https://www.elastic.co/guide/en/beats/filebeat/current/kafka-output.html#_partition

Right now partition is the following:

###### random

```
    partition: random
    random:
      group_events: 1
```

So this should look like

```
    partition:
      random:
        group_events: 1
```

`partition` should be an object, i.e. `partition.random.group_events: 1`

###### round_robin

Similarly, for `round_robin`

Right now it looks like

```
    partition: round_robin
    round_robin:
      group_events: 1
```

it should be

```
    partition:
       round_robin:
         group_events: 1
```

###### hash

for hash:

```
    partition:
      hash:
        hash: 'asdasd,asdasd'
```

###### yaml generation bug?

I also think there might be a bug here:

When I change the `partition` a few times, it looks like `hash`, `round_robin`, and `random` get appended to the yaml.  We should just be replacing the entire `partition` field each time depending on the user selection.

I'm seeing this in the yaml, we should follow the format above and ensure we just have `partition.{strategy}.{group_events | hash}`

```
    partition: hash
    random:
      group_events: 1
    round_robin:
      group_events: 1
    hash:
      hash: 'asdasd,asdasd'
```

#### Topics

We need to follow the rules here: https://www.elastic.co/guide/en/beats/filebeat/current/kafka-output.html#topic-option-kafka

For topics, I'm seeing the following

```
  topics:
    - topic: asd
      when:
        type: contains
        condition: sdf
    - topic: asd
```

We should actually see this as:

```
topics:
  - topic: asd
    when:
      contains:
        message: sdf
  - topic: asd
```

- `type: contains` should a field under `when:` that is just called `contains`. 
-  `condition: sdf` should be `message: sdf` and be nested under `when`.

szwarckonrad commented:
@kilfoyle I changed the copy, however, hrefs come from `kbn-doc-links` package and we use them, following logstash example, like that `href={docLinks.links.fleet.settings}`. My guess would be that change there is required so that we can consume these links for both logstash and kafka. CC @kevinlog 
![Screenshot 2023-08-03 at 14 43 40](https://github.com/elastic/kibana/assets/29123534/76ce0d3e-1a78-475a-ad51-b2bc1aac984b)

kilfoyle commented:
> I changed the copy, however, hrefs come from kbn-doc-links package and we use them, following logstash example, like that href={docLinks.links.fleet.settings}. My guess would be that change there is required so that we can consume these links for both logstash and kafka.

Nice. Thanks a lot for fixing the copy, Konrad! Now that I understand how the hrefs are handled, I'll wait for my docs PR to merge (so that the URLs exist) and then I'll open a separate issue or PR to have the hrefs updated.

cmacknz commented:
Can we add https://github.com/elastic/kibana/issues/163093 to the changes in this PR as well?

We should remove the Channel Buffer Size configuration, it's an advanced option so can go in the Advanced YAML parameters box when needed.

kevinlog commented:
@elasticmachine merge upstream

kevinlog commented:
@szwarckonrad a few more changes after some additional testing.

#### Remove channel buffer size

We should remove the channel buffer size field entirely.

https://github.com/elastic/kibana/issues/163093

#### One more tweak to topics

Right now, we're getting the following:

```
    topics:
      - topic: asd
        contains:
          message: asd
      - topic: asd
        range:
          message: asd
      - topic: sad
```

We should be getting:
```
    topics:
      - topic: asd
        when:
          contains:
            message: asd
      - topic: asd
        when:
          range:
            message: asd
      - topic: sad
```

Note that we should have `when` above the conditions (i.e. `contains`, `range`).

#### Hosts field should accept `host:port`

The hosts field should actually only support `host:port` - we should not include the `protocol`

for instance:
- `http://localhost:1234` ❌ 
- `localhost:1234` ✅ 

Right now the validation is requiring that we enter a URL.  We should just be enforcing `host:port`.

#### The client ID field should not have any spaces

The Client ID field should NOT have any spaces.  We should change the default to `Elastic` (one word).  Also, we should add validation with this regex `^[A-Za-z0-9._-]+$`

szwarckonrad commented:
Changes:
1. Added `when` nested property to topics - 13c0a6117b63a9443e11c4b1bb1fe2e925be7312
![Screenshot 2023-08-04 at 11 59 18](https://github.com/elastic/kibana/assets/29123534/c454f8a5-e0ad-479b-9ab9-4bbbd1915249)
2. Added client_id validation - c68b35c4463ccb2ee7a3f04be81154cedcc13875
![test](https://github.com/elastic/kibana/assets/29123534/ee581aab-59cf-43b7-a82e-bdcda02ff62f)
3. Removed channel buffer size option from UI, API, YAML - 92618f028b16d43b7160d84839d3b105d4342bc9
![Screenshot 2023-08-04 at 11 57 49](https://github.com/elastic/kibana/assets/29123534/36f696fe-f521-4ed0-a5af-9f48553ecff6)
4. Changed host validation for kafka. Now it requires host:port format - 783bef612160c3f5f2cc6fd4a6ac06fb90423dd9
![test](https://github.com/elastic/kibana/assets/29123534/434e1d7c-04a3-4043-8b08-0f67ffa8fb2f)

nimarezainia commented:
@kilfoyle wanted to bring this PR to your attention. In particular this [comment](https://github.com/elastic/kibana/pull/162875#issuecomment-1664665950). Could we please include this information in the docs page also? thank you.

szwarckonrad commented:
The CA text has been relocated from the Authentication panel. It is now being displayed, along with the added Verification Mode selection below the panel, for both the `username/password`  and `ssl` options, it remains hidden for the `none` option.


https://github.com/elastic/kibana/assets/29123534/f18e4d82-c885-4163-a5b6-ca45e66ca379

szwarckonrad commented:
ef21a4a82899a7f3302af3748ecf0d168640dab2 
1. Rolled back hiding CA and Verification Mode in `None` auth scenario. Now its being shown on all three modes of authentication.
2. Removed non used `auth_type` field from yaml.
3. Empty headers will be shown as `headers: []` in yaml from now on.

kilfoyle commented:
@szwarckonrad Just wanted to say thanks for including the screen recordings! These are really helpful as I work through the associated [docs changes](https://github.com/elastic/ingest-docs/pull/336).

kevinlog commented:
@szwarckonrad @faec - I'm summarizing the additional changes that we discussed over slack to address the remaining authentication issues.

- We will add another radio button set in the UI for "Connection" that appears under "Authentication" when the user selects `none`.  "Connection" only appears under "Authentication" if the user has selected `none`
- The "Connection" field will have two values, `plaintext` or `encryption`.
   - If `plaintext` is selected, we **do not** show the CA or Verification Method fields or include it in the policy yaml
   - If `encryption` is selected, we **do** show the CA and Verification Method fields.  CA is always optional.  Verification Method will have a value that we pass.
- Username/pass and SSL options are unchanged.  They will have the option CA and Verification Method fields as they do today.

Details on yaml generation:
- `connection_type: encrypted` UI side results in yaml with `ssl: {ca, verification_mode}` , `ca` being optional, `verification_mode` will always be there.
- `connection_type: plaintext` UI side results in yaml without `ssl:{}`
- For `user_pass` and `ssl` nothing changes - both will always have `ssl: {verification_method}` and optional `ssl.ca`
- `connection_type` will be kept in UI/API context only, wont be passed to the .yaml if not needed there

nchaulet commented on x-pack/plugins/fleet/server/saved_objects/index.ts:
```diff
@@ -229,6 +232,29 @@ const getSavedObjectTypes = (): { [key: string]: SavedObjectsType } => ({
         broker_timeout: { type: 'integer' },
         broker_ack_reliability: { type: 'text' },
         broker_buffer_size: { type: 'integer' },
+        required_acks: { type: 'integer' },
```
We should probably add `index: false` to all of those kafka fields as it seems to me that we will not search on that

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

szwarckonrad commented on x-pack/plugins/fleet/public/applications/fleet/sections/settings/components/edit_output_flyout/output_form_kafka.tsx:
```diff
@@ -84,13 +84,13 @@ export const OutputFormKafkaSection: React.FunctionComponent<Props> = (props) =>
+            defaultMessage="Specify the URLs that your agents will use to connect to Kafka. {guideLink}."
             values={{
               guideLink: (
                 <EuiLink href={docLinks.links.fleet.settings} target="_blank" external>
```
https://github.com/elastic/kibana/pull/162875#issuecomment-1664010559
:)

juliaElastic commented on x-pack/plugins/fleet/server/saved_objects/migrations/to_v8_10_0.ts:
could we add a unit test on the migration?

juliaElastic reviewed (approved):
LGTM, please check comments.

gsoldevila reviewed (approved):
LGTM changes in the `ingest-outputs` mappings are compatible (non-breaking).

nchaulet commented:
I think there is one part we may been missing for Kafka output (this should probably be a follow up PR) it's the whole preconfiguration thing to be able to define a kafka output in kibana config file, probably need to update that function and check the schema support it https://github.com/elastic/kibana/blob/40db94ae4d1a1a48e9833eb59b3dd145d1343d7d/x-pack/plugins/fleet/server/services/preconfiguration/outputs.ts#L167
